vector-stores/spring-ai-opensearch-store/src/main/java/org/springframework/ai/vectorstore/OpenSearchVectorStore.java

```diff
@@ -0,0 +1,202 @@
+                .query(getOpenSearchSimilarityQuery(embedding, filterExpression))
+                .size(topK)
+                .minScore(similarityThreshold)
+                .build());
+    }
+
+    private Query getOpenSearchSimilarityQuery(List<Double> embedding, Filter.Expression filterExpression) {
+        return Query.of(queryBuilder -> queryBuilder.scriptScore(scriptScoreQueryBuilder -> {
+            scriptScoreQueryBuilder.query(
+                            queryBuilder2 -> queryBuilder2.queryString(queryStringQuerybuilder -> queryStringQuerybuilder
+                                    .query(getOpenSearchQueryString(filterExpression))))
+                    .script(scriptBuilder -> scriptBuilder
+                            .inline(inlineScriptBuilder -> inlineScriptBuilder.source("knn_score")
+                                    .lang("knn")
+                                    .params("field", JsonData.of("embedding"))
+                                    .params("query_value", JsonData.of(embedding))
+                                    .params("space_type", JsonData.of(this.similarityFunction))));
+            // https://opensearch.org/docs/latest/search-plugins/knn/knn-score-script
+            // k-NN ensures non-negative scores by adding 1 to cosine similarity, extending OpenSearch scores to 0-2.
+            // A 0.5 boost normalizes to 0-1.
+            return this.similarityFunction.equals(COSINE_SIMILARITY_FUNCTION) ? scriptScoreQueryBuilder.boost(
+                    0.5f) : scriptScoreQueryBuilder;
+        }));
+    }
+
+    private String getOpenSearchQueryString(Filter.Expression filterExpression) {
+        return Objects.isNull(filterExpression) ? "*"
+                : this.filterExpressionConverter.convertExpression(filterExpression);
+
+    }
+
+    private List<Document> similaritySearch(org.opensearch.client.opensearch.core.SearchRequest searchRequest) {
+        try {
+            return this.openSearchClient.search(searchRequest, Document.class)
+                    .hits()
+                    .hits()
+                    .stream()
+                    .map(this::toDocument)
+                    .collect(Collectors.toList());
+        } catch (IOException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    private Document toDocument(Hit<Document> hit) {
+        Document document = hit.source();
+        document.getMetadata().put("distance", 1 - hit.score().floatValue());
+        return document;
+    }
+
+    public boolean exists(String targetIndex) {
+        try {
+            BooleanResponse response = this.openSearchClient.indices()
+                    .exists(existRequestBuilder -> existRequestBuilder.index(targetIndex));
+            return response.value();
+        } catch (IOException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    public CreateIndexResponse createIndexMapping(String index, Map<String, Property> properties) {
```

Comment:
is this supposed to be public?

Reply:
Yes, It's public to handle different sizes for each embedding model and allow for custom metadata mappings.

Reply:
https://github.com/spring-projects/spring-ai/pull/508/commits/7b77b193c0bc4cc54abf49041b932725ca268568

Refer to the commit where I added the mappingJson parameter to OpenSearchVectorStore to handle different embedding sizes and allow custom metadata mappings.